thunder/dev_utils/export_stateful_ex_transform.py

```diff
@@ -0,0 +1,123 @@
+        containers for references and implements __call__(...) to resolve
+
+    Usage:
+    1) Register once at import/init time. For example, for TransformerEngine:
+       ExportStatefulExecutorsTransform.register_ref_callback(
+           "te_fp8_states", register_cb, resolve_cb, StatsClass
+       )
+    2) Enable at compile time:
+       thunder.jit(model, executors=[...], transforms=[..., ExportStatefulExecutorsTransform()])
+    3) After each run, call `module.te_fp8_states()` to resolve and return the latest values.
+
+    Notes:
+    - Supports multiple ThunderModule instances (e.g., subgraphs)
+    - Callback errors are swallowed to avoid interfering with execution
+    """
+
+    _register_callbacks: dict[str, Callable] = {}
+    _callback_attributes: list[tuple[str, type[ExportStatefulExecutorsStats], Callable]] = []
+
+    _instance = None
+
+    def __new__(cls, *args, **kwargs):
+        """Ensure singleton instance across repeated transform construction."""
+        if cls._instance is None:
+            cls._instance = super().__new__(cls)
+        return cls._instance
+
+    def __init__(self):
+        """Initialize internal weakrefs registry.
+
+        ThunderCompiler and other compilation flows may create multiple
+        `ThunderModule` instances for subgraphs; we keep weak references
+        to update all of them during post-optimization registration.
+        """
+        self.tm_refs = []
+
+    @classmethod
+    def register_ref_callback(
+        cls, name: str, callback: Callable, resolve_cb: Callable, instance: type[ExportStatefulExecutorsStats]
+    ) -> None:
+        """Register per-executor reference and resolver callbacks.
+
+        Installs a module attribute named `name` by constructing `instance` with
+        the resolver function. The `callback` will be invoked during
+        post-optimization to record reference locations on the module.
+
+        Args:
+            name: Module attribute to attach (e.g., "te_fp8_states").
+            callback: Function `(trace, module) -> None` that records refs.
+            resolve_cb: Function `(module) -> Any` that resolves values on demand.
+            instance: A class (must be a subclass of ExportStatefulExecutorsStats) constructed as `instance(module, resolve_cb)`.
+        """
+        if not issubclass(instance, ExportStatefulExecutorsStats):
+            raise TypeError(f"Provided instance {instance} must be a subclass of ExportStatefulExecutorsStats")
+        cls._register_callbacks[name] = callback
+        cls._callback_attributes.append((name, instance, resolve_cb))
+
+    def transform_module(self, model) -> None:
+        assert model is not None
+        # Cache a weakref to the ThunderModule for later runtime export
+        self.tm_refs.append(weakref.ref(model))
+        # Initialize attributes on model
+        for name, instance, resolve_cb in self._callback_attributes:
+            setattr(model, name, instance(model, resolve_cb))
```

Comment:
I think this line leads to reference cycle -> model holds a reference to instance and instance holds a reference to the model.